xarray/core/variable.py

```diff
@@ -406,6 +407,96 @@ def __getitem__(self, key):
         return type(self)(dims, values, self._attrs, self._encoding,
                           fastpath=True)
 
+    def _broadcast_indexes(self, key):
+        """
+        Parameters
+        -----------
+        key: One of
+            array
+            a mapping of dimension names to index.
+
+        Returns
+        -------
+        dims: Tuple of strings.
+            Dimension of the resultant variable.
+        indexers: list of integer, array-like, or slice. This is aligned
+            along self.dims.
+        """
+        if not utils.is_dict_like(key):
+            key = {self.dims[0]: key}
+        example_v = None
+        indexes = OrderedDict()
+        for k, v in key.items():
+            if not isinstance(v, (integer_types, slice, Variable)):
+                if not hasattr(key, 'ndim'):  # convert list or tuple
+                    v = np.array(v)
+            if example_v is None and isinstance(v, Variable):
+                example_v = v
+            indexes[k] = v
+
+        # When all the keys are array or integer, slice
```

Comment:
I think we actually need two totally different code paths for [basic vs. advanced indexing](https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html):
- If all indexers are integers/slices, we can use NumPy's "basic indexing". There's no need to futz with broadcasting. We can use exactly a simpler version of the existing logic, dropping _`indexable_data`. (We need this path because "basic indexing" is much faster than "advanced indexing" in NumPy, as it avoids copying data.)
- If any indexers are arrays or Variable objects, we need to do "advanced indexing":
    - Normalize every `(key, value)` pair into `xarray.Variable` objects:
        - `xarray.Variable` value -> keep them unmodified
        - Integer values -> `xarray.Variable((), value)`
        - Slice values -> `xarray.Variable((key,), np.arange(*value.indexes(size)))` where `size` is the size of the dimension corresponding to `key`.
        - 1D list/numpy arrays -> `xarray.Variable((key,), value)`
        - N-D arrays -> IndexError
    - Make all indexers broadcast compatible with a single call to `_broadcast_compat_variables()`. All variables are now guaranteed to have the same `dims`, so take `dims` from the first variable to label the result of indexing.
    - Create a tuple using `data` from each of these variables, and use it to index this variable's data.